src/server.py

```diff
@@ -33,94 +68,155 @@ def server():
-            request = b''.join(request)
-            print('Request received:')
-            print(request.decode('utf-8'))
-            try:
-                uri = parse_request(request)
-                response = response_ok()
-            except ValueError as e:
-                response = response_error(*e.args)
-            conn.sendall(response)
-            conn.close()
-            time.sleep(0.01)
-    except KeyboardInterrupt:
-        pass
+            http_server(conn, addr)
+    except Exception as e:
+        print(e.msg)
     finally:
-        try:
-            conn.close()
-        except NameError:
-            pass
         print('\nShutting down the server...\n')
         serv_sock.close()
         sys.exit()
 
 
-def parse_request(request):
-    """TOOD: Docstring."""
-    # only accept GET
-    # only HTTP/1.1
-    # validate that proper Host: header was specified
-    # other requests raise appropriate Python error
-    # if no conditions arise, should return the URI
-    method = b''
-    uri = b''
-    protocol = b''
-    headers = b''
+def http_server(conn, addr):
+    """Take HTTP requests and return appropriate HTTP response."""
+    with closing(conn):
+        request_parts = []
+        while True:
+            part = conn.recv(BUFFER_LENGTH)
+            request_parts.append(part)
+            if len(part) < BUFFER_LENGTH:
+                break
+        # Immediately decode all of incoming request into unicode.
+        # In future, may need to check Content-type of incoming request?
+        request = b''.join(request_parts).decode('utf-8')
+        print('Request received:\n{}'.format(request))
+        try:
+            uri = parse_request(request)
+            # Here body might be a bytestring.
+            body, content_type = resolve_uri(uri)
+            body_length = len(body)
+            response_headers = response_ok(content_type, body_length)
 
-    try:
-        headers, body = request.split(b'\r\n\r\n')
-    except ValueError:
-        raise ValueError(ERR_400)
+        except ValueError as e:
+            err_code = e.args[0]
+            response_headers = response_error(err_code)
+            body = HTTP_CODES[err_code].encode('utf-8')
+
+        # Re-encode into bytes on the way out.
+        response_headers = response_headers.encode('utf-8')
+
+        response = b''.join([response_headers, body])
+
+        conn.sendall(response)
+        conn.close()
+        time.sleep(0.01)
 
-    try:
-        headers = headers.split(b'\r\n')
-        first_line = headers[0]
-    except (ValueError, IndexError):
-        raise ValueError(ERR_400)
 
+def parse_request(request):
+    """Parse client request."""
+    method = ''
+    uri = ''
+    protocol = ''
+    headers = ''
+
+    # I was able to simplify this into one try/except block.
+    # The key was to organize it by error type.
     try:
+        headers, body = request.split(CRLF * 2)
+        headers = headers.split(CRLF)
```

Comment:
I really like this approach.  Split the headers away from the body, then divide up the headers into lines.